### PR TITLE
fixed the split error for LoongArch64.

### PR DESCRIPTION
**Pull request by shushanhf**
Fixed the split error for LoongArch64. https://github.com/dotnet/runtime/pull/101288#pullrequestreview-2025621422

### CODE REVIEW

**Review by tomeksowi — APPROVED** [review]

File: src/coreclr/jit/codegencommon.cpp [hunk 1]
```diff
@@ -4159,8 +4159,16 @@ void CodeGen::genHomeStackSegment(unsigned                 lclNum,
...
+    int loadOffset;
+    if (isFramePointerUsed())
+    {
+        loadOffset = -genCallerSPtoFPdelta();
+    }
+    else
+    {
+        loadOffset = -(int)seg.GetStackOffset() - genCallerSPtoInitialSPdelta();
+    }
+
```
Comment (tomeksowi) [inline src/coreclr/jit/codegencommon.cpp:4171]:
```suggestion
    int loadOffset = isFramePointerUsed() ? -genCallerSPtoFPdelta() : -genCallerSPtoInitialSPdelta();
    loadOffset += (int)seg.GetStackOffset();
```
seg.GetStackOffset() should be added to both sides, ternaries in general are good for such things. This suggestion should now fit within [coding guidelines](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/clr-jit-coding-conventions.md#1519-ternary-operators).

Comment (shushanhf) [reply src/coreclr/jit/codegencommon.cpp]: > seg.GetStackOffset() should be added to both sides, ternaries in general are good for such things. 

Thanks!

> This suggestion should now fit within [coding guidelines](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/clr-jit-coding-conventions.md#1519-ternary-operators).

```
Wrong:

(a == b) ? Foo() : Bar(); // top-level ?: disallowed
```

Comment (tomeksowi) [reply src/coreclr/jit/codegencommon.cpp]: Acceptable:
```
x = (a == b) ? 7 : 9;
```
By "top-level ?:" they mean used as an if replacement, it's ok to use in assignments.

Comment (shushanhf) [reply src/coreclr/jit/codegencommon.cpp]: I thought that the `value1` was not function's return within `condistion ? value1 : value2`.

If needed, I will update later.

File: src/coreclr/jit/abi.cpp [hunk 2]
```diff
@@ -258,16 +258,12 @@ bool ABIPassingInformation::HasExactlyOneStackSegment() const
...
+    if (NumSegments != 2)
     {
-        if (isFirstInReg != Segments[i].IsPassedInRegister())
-            return true;
+        return false;
     }
-    return false;
+
+    return Segments[0].IsPassedInRegister() && Segments[1].IsPassedOnStack();
 }
```
Comment (jakobbotsch) [inline src/coreclr/jit/abi.cpp:267]: This breaks the check for other ABIs. Please revert this change.

Comment (shushanhf) [reply src/coreclr/jit/abi.cpp:267]: OK.
But now there is only here that used the `IsSplitAcrossRegistersAndStack()`.

Comment (shushanhf) [reply src/coreclr/jit/abi.cpp:267]: How about this ?
```
    if (NumSegments < 2)
    {   
        return false;
    }   

    bool isFirstInReg = Segments[0].IsPassedInRegister();
    for (unsigned i = 1; i < NumSegments; i++)
    {   
        if (isFirstInReg && Segments[i].IsPassedOnStack())
        {
            return true;
        }
    }
    return false;
```

I will revert it tomorrow.

Comment (jakobbotsch) [reply src/coreclr/jit/abi.cpp:267]: That would return wrong answers for a stack segment followed by a register segment.

What is the reason for changing this function exactly? The implementation looks correct to me. It looked correct to me before #101288, where it was also churned. The ABI handling code is not hot whatsoever as I have seen from our actual throughput data when working on it, so the churn does not seem justified.

Comment (shushanhf) [reply src/coreclr/jit/abi.cpp:267]: First of all, I will revert `ABIPassingInformation::IsSplitAcrossRegistersAndStack()` if needed.
I am just curious.
The `ABIPassingInformation::IsSplitAcrossRegistersAndStack()` is only used by RV64/LA64 now.

> That would return wrong answers for a stack segment followed by a register segment.

How it affect the other ABIs?

> What is the reason for changing this function exactly? The implementation looks correct to me. It looked correct to me before #101288, where it was also churned. 

before 101288 it was wrong for LA64/RV64.

**Review by tomeksowi — APPROVED** [review]

File: src/coreclr/jit/abi.cpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/abi.cpp]: > The ABIPassingInformation::IsSplitAcrossRegistersAndStack() is only used by RV64/LA64 now.
> How does it affect the other ABIs?
> This is not exist for LA64/RV64.

The ABI representation is about all of our targets. It is not only about LA64/RV64. It is meant to be general enough to handle everything we want to be able to represent. I am working on unifying all ABI handling in the JIT on it.

Basic primitive queries like `IsSplitAcrossRegistersAndStack` should be fully functional. There simply is no need for it to be partially broken just because it is only being used by LA64/RV64 at this point in time.

> before 101288 it was wrong for LA64/RV64.

How was `ABIPassingInformation::IsSplitAcrossRegistersAndStack` wrong before?

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/abi.cpp (hunk 2 above)
Comment (tomeksowi) [reply src/coreclr/jit/abi.cpp]: > bool anyReg   = false;
>     bool anyStack = false;
>     for (unsigned i = 0; i < NumSegments; i++)
>     {
>         anyReg |= Segments[i].IsPassedInRegister();  // this is `i`
>         anyStack |= Segments[i].IsPassedOnStack(); // this is also `i`  ,  then the `anyReg && anyStack` is always false.
>     }
>     return anyReg && anyStack;

It was a bit redundant (IsPassed{InRegister,OnStack} are mutually exclusive) and iterated unnecessarily till the end (doesn't really matter, there's so few segments) so I did a minor cleanup. But IMO it was correct, it returned true if at least one segment was in register and at least one on stack.

Comment (jakobbotsch) [reply src/coreclr/jit/abi.cpp]: > BTW, you can delete the abi's arg-info within the `Compiler::lvaInitUserArgs`.

I will remove it once all uses of it have been removed. There are still uses of `LclVarDsc::GetArgReg()`, the HFA related functions, etc.

> > How was `ABIPassingInformation::IsSplitAcrossRegistersAndStack` wrong before?
> 
> ```
>     bool anyReg   = false;
>     bool anyStack = false;
>     for (unsigned i = 0; i < NumSegments; i++)
>     {
>         anyReg |= Segments[i].IsPassedInRegister();  // this is `i`
>         anyStack |= Segments[i].IsPassedOnStack(); // this is also `i`  ,  then the `anyReg && anyStack` is always false.
>     }
>     return anyReg && anyStack;
> ```

`anyReg` is true if any segment is passed in register. `anyStack` is true if any segment is passed on stack. Both can be true if there are two segments where one is passed in register and the other passed on stack.

Comment (shushanhf) [reply src/coreclr/jit/abi.cpp]: > But IMO it was correct, it returned true if at least one segment was in register and at least one on stack.

Yes, you are right.
